mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h

```diff
@@ -6447,79 +7055,152 @@ struct GridwiseGemmV2RewritePattern
+    // llvm::errs() << GemmABlockCopyThreadSliceLengths_GemmM << " ";
+    // llvm::errs() << GemmABlockCopyThreadSliceLengths_GemmKPack << "\n";
+
+    if (GemmABlockCopyThreadSliceLengths_GemmK == 0 ||
+        GemmABlockCopyThreadSliceLengths_GemmM == 0 ||
+        GemmABlockCopyThreadSliceLengths_GemmKPack == 0) {
+      llvm::errs() << "Blockwise copy slice lengths for matrix A is zero which "
+                      "is invalid.\n";
+      return failure();
+    }
 
     // Compute ThreadClusterLengths for Matrix A.
+    uint64_t GemmABlockCopyClusterLengths_GemmKPack =
+        KPack / GemmABlockCopyThreadSliceLengths_GemmKPack;
     uint64_t GemmABlockCopyClusterLengths_GemmK =
         KPerBlock / GemmABlockCopyThreadSliceLengths_GemmK;
-    // int64_t GemmABlockCopyClusterLengths_GemmM =
+    // uint64_t GemmABlockCopyClusterLengths_GemmM =
     //    MPerBlock / GemmABlockCopyThreadSliceLengths_GemmM;
 
     // llvm::errs() << "thread cluster lengths for Matrix A\n";
     // llvm::errs() << GemmABlockCopyClusterLengths_GemmK << " ";
-    // llvm::errs() << GemmABlockCopyClusterLengths_GemmM << "\n";
+    // llvm::errs() << GemmABlockCopyClusterLengths_GemmM << " ";
+    // llvm::errs() << GemmABlockCopyClusterLengths_GemmKPack << "\n";
 
     // Compute ThreadSliceLengths for Matrix B.
     uint64_t GemmBBlockCopyNumberDataPerThread =
-        NPerBlock * KPerBlock / BlockSize;
+        NPerBlock * KPerBlock * KPack / BlockSize;
+
+    // llvm::errs() << "GemmBBlockCopyNumberDataPerThread: "
+    //              << GemmBBlockCopyNumberDataPerThread << "\n";
 
     uint64_t GemmBBlockCopyThreadSliceLengths_GemmK;
     uint64_t GemmBBlockCopyThreadSliceLengths_GemmN;
+    uint64_t GemmBBlockCopyThreadSliceLengths_GemmKPack = 1;
     switch (matrix_b_source_vector_read_dim) {
     case GemmK:
-      GemmBBlockCopyThreadSliceLengths_GemmK = matrix_b_source_data_per_read;
-      GemmBBlockCopyThreadSliceLengths_GemmN =
-          GemmBBlockCopyNumberDataPerThread /
-          GemmBBlockCopyThreadSliceLengths_GemmK;
+      if (KPack > 1) {
+        GemmBBlockCopyThreadSliceLengths_GemmKPack =
+            matrix_b_source_data_per_read;
+        GemmBBlockCopyThreadSliceLengths_GemmK =
+            KPack / matrix_b_source_data_per_read;
+        GemmBBlockCopyThreadSliceLengths_GemmN =
+            GemmBBlockCopyNumberDataPerThread / KPack;
+      } else {
+        GemmBBlockCopyThreadSliceLengths_GemmK = matrix_b_source_data_per_read;
+        GemmBBlockCopyThreadSliceLengths_GemmN =
+            GemmBBlockCopyNumberDataPerThread /
+            GemmBBlockCopyThreadSliceLengths_GemmK;
+      }
       break;
     case GemmMorN:
-      GemmBBlockCopyThreadSliceLengths_GemmN = matrix_b_source_data_per_read;
-      GemmBBlockCopyThreadSliceLengths_GemmK =
-          GemmBBlockCopyNumberDataPerThread /
-          GemmBBlockCopyThreadSliceLengths_GemmN;
+      // TBD: FIXME. Review logic here.
+      if (KPack > 1) {
```

Comment:
As a general thought, would anything break (especially given #440) if, instead of all these `if KPack > 1` checks, we just always had a KPack dimension that was sometimes fixed to 1?

Reply:
(and yeah, I know that applies to #441's approach to the output tensor as well)